### ISSUE TITLE
Runtime TileData Z-Index modifications are reset after TileMapLayer _rendering_update

### ISSUE THREAD

**Issue by Shiva-Shadowsong** [godotengine/godot#93327]
### Tested versions

- Reproducible in v4.3.beta1.official [a4f2ea91a]

### System information

Godot v4.3.beta1 - Windows 10.0.22631 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1660 SUPER (NVIDIA; 31.0.15.2802) - AMD Ryzen 5 5600X 6-Core Processor (12 Threads)

### Issue description

If I modify the `z_index` of tiles in a `TileMapLayer` in `_tile_data_runtime_update`, that z_index modification is gone if the TileMapLayer re-renders. 

I noticed it in my floor system, because I `hide()` the floor when it shouldn't be visible, and `show()` it again when it is. Upon being shown again, the tiles I applied a z_index modification to get reverted back to the initial state from their original `TileData`.

![1Ln1Y9o 1](https://github.com/godotengine/godot/assets/15309863/232f3ead-075a-426e-83e2-22bab4b1f276)

## Expectation

I would expect that if I modify a z_index (or any other property of a cell), the change would remain cached and not reset unless I explicitly set it to some other value or reset it.

## Notable Mention

It's also maybe important/related to mention that if you try to `get_cell_tile_data(coords)` on a cell you've modified in runtime, you still get the original `TileData`, and you have no way to access the modifications you made unless you explicitly save them in some other data structure.

I understand that maybe `get_cell_tile_data` is *supposed* to get the original `TileData` object, but there could be a way added to get the cached runtime tiledata with changes too. This may be a candidate for a different report though.

### Steps to reproduce

1. In `_use_tile_data_runtime_update` of a TileMapLayer, specify some coords that require an update.
2. In `_tile_data_runtime_update`, for those coordinates, apply a `z_index` update (In my case, I add +10 z_index).
3. `hide()` that `TileMapLayer`, then `show()` it again. Your `z_index` change will be reset to its initial value.

### Minimal reproduction project (MRP)

[mrp_tiledatabug.zip](https://github.com/user-attachments/files/15892847/mrp_tiledatabug.zip)

**Comment by KoBeWi** [comment]
AFAIK runtime update runs every time the TileMapLayer is updated, so you need to apply your changes every time. Does it happen with other properties?

**Comment by Shiva-Shadowsong** [comment]
> AFAIK runtime update runs every time the TileMapLayer is updated, so you need to apply your changes every time. Does it happen with other properties?

Yes, for example here is the result when I also update the modulate alongside z_index.

![https://i.imgur.com/nU6oxmK.gif](https://i.imgur.com/nU6oxmK.gif)

**Comment by KoBeWi** [comment]
The way runtime updates work now is that the user receives a copy of TileData, which is then used to build tiles. However the modified data is not stored anywhere, you always get a new copy. Thus you have to update the modifications each time.

**Comment by KoBeWi** [comment]
> If I modify a tile with _tile_data_runtime_update, do I have to keep updating it until the end of time so as not to lose the modification?

Yes.

**Comment by TheKassaK** [comment]
> > If I modify a tile with _tile_data_runtime_update, do I have to keep updating it until the end of time so as not to lose the modification?
> 
> Yes.

That makes it clear :p Thanks for the quick reply.

I don't understand the point of all this then, if from the moment you modify a tile in this way you commit to having to update it ad infinitum, the impact on performance is just huge. In my case, I have to change the material (shader), modify the physics or the occlusion layer...

I'll find another way then.

**Comment by KoBeWi** [comment]
That doesn't mean the behavior has to stay this way. We are discussing a potential solution: https://chat.godotengine.org/channel/2d
